services/server/src/server/services/storageServices/EtherscanVerifyAPIService.ts

```diff
@@ -0,0 +1,586 @@
+    sourcifyDatabaseService: SourcifyDatabaseService,
+    options: EtherscanVerifyAPIServiceOptions,
+  ) {
+    this.IDENTIFIER = identifier;
+    this.database = sourcifyDatabaseService.database;
+    this.options = {
+      chainApiUrls: options.chainApiUrls || {},
+      apiKeys: options.apiKeys || {},
+      defaultApiKey: options.defaultApiKey || "",
+    };
+  }
+
+  async init(): Promise<boolean> {
+    if (Object.keys(this.options.chainApiUrls).length === 0) {
+      switch (this.IDENTIFIER) {
+        case WStorageIdentifiers.EtherscanVerify:
+          this.options.chainApiUrls = await fetchEtherscanChainApiUrls();
+          break;
+        case WStorageIdentifiers.BlockscoutVerify:
+          this.options.chainApiUrls = await fetchBlockscoutChainApiUrls();
+          break;
+        case WStorageIdentifiers.RoutescanVerify:
+          this.options.chainApiUrls = await fetchRoutescanChainApiUrls();
+          break;
+        default:
+          throw new Error(`Unsupported verifier: ${this.IDENTIFIER}`);
+      }
+    }
+    return true;
+  }
+
+  async storeVerification(
+    verification: VerificationExport,
+    jobData?: {
+      verificationId: string;
+      finishTime: Date;
+    },
+  ): Promise<void> {
+    if (
+      verification.compilation.language === "Vyper" &&
+      (this.IDENTIFIER === WStorageIdentifiers.RoutescanVerify ||
+        this.IDENTIFIER === WStorageIdentifiers.BlockscoutVerify)
+    ) {
+      logger.info(
+        `${this.IDENTIFIER} does not support Etherscan API Vyper contract verification`,
+        {
+          chainId: verification.chainId,
+          address: verification.address,
+        },
+      );
+      throw new Error(
+        `${this.IDENTIFIER} does not support Etherscan API Vyper contract verification`,
+      );
+    }
+
+    const submissionContext = {
+      identifier: this.IDENTIFIER,
+      chainId: verification.chainId,
+      address: verification.address,
+    };
+
+    const apiBaseUrl = this.getApiBaseUrl(verification.chainId);
+    if (!apiBaseUrl) {
+      logger.warn(
+        "No Etherscan API endpoint configured for chain",
+        submissionContext,
+      );
+      return;
+    }
+
+    const response = await this.submitVerification(verification, apiBaseUrl);
+
+    logger.info("Submitted verification to explorer", {
+      ...submissionContext,
+      receiptId: response.result,
+    });
+
+    if (!jobData?.verificationId) {
+      return;
+    }
+
+    // Record the result of the failed submission by storing the error message
+    if (response.status !== "1" || response.message !== "OK") {
+      logger.warn("Explorer rejected verification submission", {
+        ...submissionContext,
+        verificationJobId: jobData.verificationId,
+        error: response.result,
+      });
+      await this.database.upsertExternalVerification(
+        jobData.verificationId,
+        this.IDENTIFIER,
+        {
+          error: response.result,
+        },
+      );
```

Comment:
This call should also be wrapped in a try catch for consistency. Could also be moved to a private function to reduce code duplication.